core/src/lib.rs

```diff
@@ -89,39 +96,65 @@ pub fn init_worker<CT>(
 where
     CT: Into<sealed::AnyClient>,
 {
-    let client = init_worker_client(&worker_config, *client.into().into_inner());
-    if client.namespace() != worker_config.namespace {
+    let client_inner = *client.into().into_inner();
+    let endpoint = client_inner.options().clone().target_url;
+    let client = init_worker_client(
+        worker_config.namespace.clone(),
+        worker_config.client_identity_override.clone(),
+        client_inner,
+    );
+    let namespace = worker_config.namespace.clone();
+    if client.namespace() != namespace {
         bail!("Passed in client is not bound to the same namespace as the worker");
     }
     if client.namespace() == "" {
         bail!("Client namespace cannot be empty");
     }
     let client_ident = client.get_identity().to_owned();
-    let sticky_q = sticky_q_name_for_worker(&client_ident, &worker_config);
+    let sticky_q = sticky_q_name_for_worker(&client_ident, worker_config.max_cached_workflows);
 
     if client_ident.is_empty() {
         bail!("Client identity cannot be empty. Either lang or user should be setting this value");
     }
 
-    let heartbeat_fn = worker_config
-        .heartbeat_interval
-        .map(|_| Arc::new(OnceLock::new()));
-
     let client_bag = Arc::new(WorkerClientBag::new(
         client,
-        worker_config.namespace.clone(),
+        namespace.clone(),
         client_ident,
         worker_config.versioning_strategy.clone(),
-        heartbeat_fn.clone(),
     ));
 
-    Ok(Worker::new(
+    let mut worker = Worker::new(
         worker_config,
         sticky_q,
-        client_bag,
+        client_bag.clone(),
         Some(&runtime.telemetry),
-        heartbeat_fn,
-    ))
+        false,
+    );
+
+    if let Some(_) = runtime.heartbeat_interval {
+        let worker_instance_key = worker.worker_instance_key();
+        let heartbeat_callback = worker
+            .get_heartbeat_callback()
+            .expect("Worker heartbeat data should exist for non-shared namespace worker");
```

Comment:
I'd probably make this a dbg panic that just skips registration if it's not set.